src/refiners/foundationals/latent_diffusion/freeu.py

```diff
@@ -0,0 +1,110 @@
+
+
+# Ported from https://github.com/huggingface/diffusers/blob/main/src/diffusers/utils/torch_utils.py
+def fourier_filter(x_in: Tensor, threshold: int, scale: float) -> Tensor:
+    """Fourier filter as introduced in FreeU (https://arxiv.org/abs/2309.11497).
+
+    This version of the method comes from here:
+    https://github.com/huggingface/diffusers/pull/5164#issuecomment-1732638706
+    """
+    x = x_in
+    B, C, H, W = x.shape
+
+    # Non-power of 2 images must be float32
+    if (W & (W - 1)) != 0 or (H & (H - 1)) != 0:
+        x = x.to(dtype=torch.float32)
+
+    # FFT
+    x_freq = fftn(x, dim=(-2, -1))  # type: ignore
+    x_freq = fftshift(x_freq, dim=(-2, -1))  # type: ignore
+
+    B, C, H, W = x_freq.shape  # type: ignore
+    mask = torch.ones((B, C, H, W), device=x.device)  # type: ignore
+
+    crow, ccol = H // 2, W // 2  # type: ignore
+    mask[..., crow - threshold : crow + threshold, ccol - threshold : ccol + threshold] = scale
+    x_freq = x_freq * mask  # type: ignore
+
+    # IFFT
+    x_freq = ifftshift(x_freq, dim=(-2, -1))  # type: ignore
+    x_filtered = ifftn(x_freq, dim=(-2, -1)).real  # type: ignore
+
+    return x_filtered.to(dtype=x_in.dtype)  # type: ignore
+
+
+# Ported from https://github.com/huggingface/diffusers/blob/main/src/diffusers/utils/torch_utils.py
+class FreeUHiddenStates(Module):
+    def __init__(self, b: float):
+        super().__init__()
+        self.b = b
+
+    def forward(self, hidden_states: Tensor) -> Tensor:
+        num_half_channels = hidden_states.shape[1] // 2
+        hidden_states[:, :num_half_channels] = hidden_states[:, :num_half_channels] * self.b
+        return hidden_states
+
+
+class FreeUSkipConnections(Module):
+    def __init__(self, s: float):
+        super().__init__()
+        self.s = s
+
+    def forward(self, hidden_states: Tensor) -> Tensor:
+        hidden_states = fourier_filter(hidden_states, threshold=1, scale=self.s)
+        return hidden_states
+
+
+class FreeUResidualConcatenator(fl.Chain):
+    def __init__(self, n: int, b: float, s: float) -> None:
+        self.n = n
+
+        super().__init__(
+            fl.Concatenate(
+                FreeUHiddenStates(b),
+                fl.Chain(
+                    fl.UseContext(context="unet", key="residuals").compose(lambda residuals: residuals[self.n]),
+                    FreeUSkipConnections(s),
+                ),
+                dim=1,
+            ),
+        )
+
+
+class SDFreeUAdapter(Generic[T], fl.Chain, Adapter[T]):
+    def __init__(self, target: T, bs: list[float], ss: list[float]):
+        assert len(bs) == len(ss)
+        self.bs = bs
+        self.ss = ss
+        with self.setup_adapter(target):
+            super().__init__(target)
+
+    def inject(self: TSDFreeUAdapter, parent: fl.Chain | None = None) -> TSDFreeUAdapter:
+        for n, block in enumerate(cast(Iterable[fl.Chain], self.target.UpBlocks)):
+            if n >= len(self.bs):
```

Comment:
instead of iteratiing through the blocks, iterate on enumerate(zip(bs, ss)) so you don't need a break